### PR TITLE
Allow to set same mode as existing if load fails.

### PR DESCRIPTION
**Pull request by Carreau**
Setting the mode many times in a row, would fail to actually set the
mode if it was failing to load the mode the first time.

The prevent the above behavior by falling back **explicitly** to 'null'
if mode does not exist yet, or can't be loaded.

---

I think it would be nice to backport that for IHaskell. ping @gibiansky.

### CODE REVIEW

**Comment by minrk** [comment]
:+1:

File: IPython/html/static/notebook/js/notebook.js [hunk 1]
```diff
@@ -1495,29 +1495,42 @@ define(function (require) {
...
-                    // This is currently redundant, because cm_config ends up as
-                    // codemirror's own .options object, but I don't want to
-                    // rely on that.
-                    cell.cm_config.mode = spec;
-                }
-            });
+            that._dispatch_mode(spec, newmode)
+        }, function(){
+            // on error don't dispatch the new mode as re-setting it later will not work.
+            that._dispatch_mode('null','null')
```
Comment (jdfreder) [inline IPython/html/static/notebook/js/notebook.js]: Why `'null'` here instead of `null`?

Comment (Carreau) [inline IPython/html/static/notebook/js/notebook.js]: `null` don't change cmMode. but `'null'` is a none name for codemirror mode that will just not highlight. (I tried `null`)

**Comment by Carreau** [comment]
Not completly right. has to add extra logic.

**Comment by jdfreder** [comment]
> Not completly right. has to add extra logic.

Okay, I set the `waiting-for-author` tag.  You can remove it when you get it figured out.

**Comment by Carreau** [comment]
Ok, I think that should be it, I locally use kernel.js
```
define(['require','./codemirror-ipython.js','base/js/namespace'], function(require,cmip, IPython){
    return {
        onload: function(){
            console.log('codemirror IPython', cmip)   
            IPython.notebook.set_codemirror_mode({name:'ipython', version:3})
        }
    }
})
```

And it seem to do the trick with this PR.

File: IPython/html/static/notebook/js/notebook.js [hunk 2]
```diff
@@ -1495,29 +1495,51 @@ define(function (require) {
...
-            that.get_cells().map(function(cell, i) {
-                if (cell.cell_type === 'code'){
-                    cell.code_mirror.setOption('mode', spec);
-                    // This is currently redundant, because cm_config ends up as
-                    // codemirror's own .options object, but I don't want to
-                    // rely on that.
-                    cell.cm_config.mode = spec;
-                }
-            });
+            that._dispatch_mode(spec, newmode)
```
Comment (minrk) [inline IPython/html/static/notebook/js/notebook.js]: semicolon

File: IPython/html/static/notebook/js/notebook.js [hunk 3]
```diff
@@ -1495,29 +1495,51 @@ define(function (require) {
...
+        }, function(){
+            // on error don't dispatch the new mode as re-setting it later will not work.
+            // don't either set to null mode if it has been changed in the meantime
+            if((_mode_equal(newmode, this.codemirror_mode))){
+                that._dispatch_mode('null','null')
```
Comment (minrk) [inline IPython/html/static/notebook/js/notebook.js]: semicolon

File: IPython/html/static/notebook/js/notebook.js [hunk 4]
```diff
@@ -1495,29 +1495,51 @@ define(function (require) {
...
-                    // codemirror's own .options object, but I don't want to
-                    // rely on that.
-                    cell.cm_config.mode = spec;
-                }
-            });
+            that._dispatch_mode(spec, newmode)
+        }, function(){
+            // on error don't dispatch the new mode as re-setting it later will not work.
+            // don't either set to null mode if it has been changed in the meantime
+            if((_mode_equal(newmode, this.codemirror_mode))){
```
Comment (minrk) [inline IPython/html/static/notebook/js/notebook.js]: extra parentheses

File: IPython/html/static/notebook/js/notebook.js [hunk 5]
```diff
@@ -1495,29 +1495,51 @@ define(function (require) {
...
+                // rely on that.
+                cell.cm_config.mode = spec;
+            }
+        });
+
+    }
+
+    // roughly try to check mode equality
+    var _mode_equal = function(mode1, mode2){
+        return ((mode1||{}).name||mode1)===((mode2||{}).name||mode2)
```
Comment (minrk) [inline IPython/html/static/notebook/js/notebook.js]: semicolon

File: IPython/html/static/notebook/js/notebook.js [hunk 6]
```diff
@@ -1495,29 +1495,51 @@ define(function (require) {
...
+
+    }
```
Comment (minrk) [inline IPython/html/static/notebook/js/notebook.js]: semicolon

File: IPython/html/static/notebook/js/notebook.js [hunk 7]
```diff
@@ -1495,29 +1495,51 @@ define(function (require) {
...
+            if (cell.cell_type === 'code'){
+                cell.code_mirror.setOption('mode', spec);
+                // This is currently redundant, because cm_config ends up as
+                // codemirror's own .options object, but I don't want to
+                // rely on that.
+                cell.cm_config.mode = spec;
+            }
+        });
+
+    }
```
Comment (minrk) [inline IPython/html/static/notebook/js/notebook.js]: semicolon

Comment (jdfreder) [inline IPython/html/static/notebook/js/notebook.js]: He must be programming in Typescript too much. ;)  Semicolons throw a compiler error on instance methods.

**Pull request merged**